kernel/src/vm/vmar/vm_mapping.rs

```diff
@@ -562,17 +527,20 @@ impl<R1, R2> VmarMapOptions<R1, R2> {
         self
     }
 
-    /// Creates the mapping.
+    /// Creates the mapping and adds it to the parent VMAR.
     ///
     /// All options will be checked at this point.
     ///
     /// On success, the virtual address of the new mapping is returned.
     pub fn build(self) -> Result<Vaddr> {
         self.check_options()?;
-        let parent_vmar = self.parent.0.clone();
-        let vm_mapping = VmMapping::build_mapping(self)?;
-        let map_to_addr = vm_mapping.map_to_addr();
-        parent_vmar.add_mapping(vm_mapping);
+
+        let Vmar(parent_vmar, _) = self.parent.unwrap();
+        let map_to_addr = parent_vmar.build_mapping(Self {
+            parent: None,
```

Comment:
Sorry for overlooking that the `VmarMapOptions` contains a `Vmar` before. Once the `VmarMapOptions` has been moved to the `vmar/mod.rs`, this operation will be redundant. The logics in `Vmar::build_mapping` can be put in this method directly, you can operate the `parent_vmar` here directly and do not need `Vmar::build_mapping` any more.